kernel/src/sched/stats/loadavg.rs

```diff
@@ -0,0 +1,107 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! This module implements the CPU load average calculation.
+//!
+//! Reference: <https://github.com/torvalds/linux/blob/master/kernel/sched/loadavg.c>
+
+use ostd::{arch::timer::TIMER_FREQ, sync::RwLock, timer};
+
+/// nr of bits of precision
+pub const FSHIFT: u64 = 11;
+/// 1.0 as fixed-point
+pub const FIXED_1: u64 = 1 << FSHIFT;
+/// 5 sec intervals
+pub const LOAD_FREQ: u64 = 5 * TIMER_FREQ + 1;
+/// 1/exp(5sec/1min) as fixed-point
+pub const EXP_1: u64 = 1884;
+/// 1/exp(5sec/5min)
+pub const EXP_5: u64 = 2014;
+/// 1/exp(5sec/15min)
+pub const EXP_15: u64 = 2037;
+
+/// Load average of all CPU cores.
+///
+/// The load average is calculated as an exponential moving average of the load
+/// over the last 1, 5, and 15 minutes.
+#[derive(Clone)]
+pub struct LoadAverage {
+    load: [u64; 3],
+    last_update: u64, // jiffies
+}
+
+impl LoadAverage {
+    /// Creates a new `LoadAverage`.
+    pub const fn new() -> Self {
+        Self {
+            load: [0, 0, 0],
+            last_update: 0,
+        }
+    }
+}
+
+/// The global load average.
+pub static LOAD_AVG: RwLock<LoadAverage> = RwLock::new(LoadAverage::new());
+
+/// Returns the calculated load average of the system.
+pub fn get_loadavg() -> [u64; 3] {
+    let mut load = LOAD_AVG.read().clone().load;
+
+    load[0] += FIXED_1 / 200;
+    load[1] += FIXED_1 / 200;
+    load[2] += FIXED_1 / 200;
+
+    load
+}
+
+/// Returns the calculated load average of the system as floating point numbers.
+pub fn get_loadavg_float() -> [f64; 3] {
+    let load = get_loadavg();
+
+    [
+        load_int(load[0]) as f64 + load_frac(load[0]) as f64 / 100.0,
+        load_int(load[1]) as f64 + load_frac(load[1]) as f64 / 100.0,
+        load_int(load[2]) as f64 + load_frac(load[2]) as f64 / 100.0,
+    ]
+}
+
+/// Updates the load average of the system.
+pub fn update_loadavg((newload, _): (u64, u64)) {
+    let jiffies = timer::Jiffies::elapsed().as_u64();
+
+    // Return if the load average was updated less than 5 seconds ago.
+    if jiffies < LOAD_AVG.read().last_update + LOAD_FREQ {
+        return;
```

Comment:
Most calls to `update_loadavg` will not do any real work.

```rust
    timer::register_callback(|| {
        loadavg::update_loadavg(nr_queued_and_running());
    });
```

But every call to `update_loadavg` obtains data from the scheduler, which involves locking and iterating over every runqueues. So to improve efficiency, I suggest refactoring the interface of `update_loadavg` to take a function pointer so that `nr_queued_and_running` can be called lazily.

Reply:
This is smart, I was worried about the efficiency of my implementation. Thank you :)